File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2019-2c-parcial2-masked/20260921-210220-97414/workdir/src/surveycrawler/crawler_facing.py
class CrawlerFacing:

    # instance creation

    @classmethod
    def facing(cls, a_facing_name):
        from surveycrawler.survey_crawler import SurveyCrawler

        for a_crawler_facing_class in cls.__subclasses__():
            if a_crawler_facing_class.is_for(a_facing_name):
                return a_crawler_facing_class()
        raise RuntimeError(SurveyCrawler.invalid_facing_error_description())

    # facing name

    @classmethod
    def facing_name(cls):
        raise NotImplementedError()

    # direction

    def direction(self):
        raise NotImplementedError()

    def opposite_direction(self):
        return self.direction().times(-1)

    # facing

    def turn_counter_clockwise(self, a_crawler):
        raise NotImplementedError()

    def turn_clockwise(self, a_crawler):
        raise NotImplementedError()

    # testing

    def is_facing(self, a_facing_name):
        return self.is_for(a_facing_name)

    @classmethod
    def is_for(cls, a_facing_name):
        return cls.facing_name() == a_facing_name

    # moving

    def retreat(self, a_crawler):
        a_crawler.move_towards(self.opposite_direction())

    def advance(self, a_crawler):
        a_crawler.move_towards(self.direction())

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2019-2c-parcial2-masked/20260921-214358-38088/workdir/src/main/java/surveycrawler/CrawlerFacing.java
package surveycrawler;

public abstract class CrawlerFacing {

    // instance creation

    public static CrawlerFacing facing(String aFacingName) {
        CrawlerFacing[] subclassInstances = subclassInstances();
        for (int i = 0; i < subclassInstances.length; i++) {
            if (subclassInstances[i].isFor(aFacingName)) {
                return subclassInstances[i];
            }
        }
        throw new RuntimeException(SurveyCrawler.invalidFacingErrorDescription());
    }

    private static CrawlerFacing[] subclassInstances() {
        return new CrawlerFacing[] {
            new CrawlerFacingRight(),
            new CrawlerFacingUp(),
            new CrawlerFacingDown(),
            new CrawlerFacingLeft()
        };
    }

    // facing name

    public abstract String facingName();

    // facing

    public abstract void turnCounterClockwise(SurveyCrawler aCrawler);

    public abstract void turnClockwise(SurveyCrawler aCrawler);

    // testing

    public boolean isFacing(String aFacingName) {
        return isFor(aFacingName);
    }

    public boolean isFor(String aFacingName) {
        return facingName().equals(aFacingName);
    }

    // moving

    public abstract void retreat(SurveyCrawler aCrawler);

    public abstract void advance(SurveyCrawler aCrawler);
}
